internal/dag/secret.go

```diff
@@ -50,35 +51,47 @@ func isValidSecret(secret *v1.Secret) (bool, error) {
 }
 
 func validateCertificate(data []byte) error {
-	cert, err := decodePEM(data)
-	if err != nil {
-		return err
-	}
-	_, err = x509.ParseCertificate(cert.Bytes)
-	return err
-}
-
-func validatePrivateKey(data []byte) error {
-	key, err := decodePEM(data)
-	if err != nil {
-		return nil
-	}
-	if _, err := x509.ParsePKCS1PrivateKey(key.Bytes); err == nil {
-		return nil
-	}
-	if _, err := x509.ParsePKCS8PrivateKey(key.Bytes); err == nil {
-		return nil
+	var exists bool
+	var block *pem.Block
+	for rest := data[:]; len(rest) > 0; {
+		block, rest = pem.Decode(rest)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		if !strings.HasSuffix(block.Type, "CERTIFICATE") {
+			continue
+		}
+		if _, err := x509.ParseCertificate(block.Bytes); err != nil {
+			return err
+		}
+		exists = true
 	}
-	if _, err := x509.ParseECPrivateKey(key.Bytes); err == nil {
-		return nil
+	if !exists {
+		return errors.New("failed to locate certificate")
 	}
-	return errors.New("unknown private key encoding")
+	return nil
 }
 
-func decodePEM(in []byte) (*pem.Block, error) {
-	block, _ := pem.Decode(in)
-	if block == nil {
-		return nil, errors.New("failed to parse PEM")
+func validatePrivateKey(data []byte) error {
+	var block *pem.Block
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		if !strings.HasSuffix(block.Type, "PRIVATE KEY") {
+			continue
+		}
+		if _, err := x509.ParsePKCS1PrivateKey(block.Bytes); err == nil {
```

Comment:
There is a question as to early exit or continue looking for more than one private key (and error'ing as a result)

Reply:
What does envoy require. I know stacking the intermediate ca's in a the certificate is common practice but what's the use case for having more than one private key block?

Reply:
IIRC it's possible to have multiple private keys and select the right key to serve to a client depending on the cipher suite negotiation and other TLS properties.

Reply:
I'll try to dig through the envoy source ti find what they accept, right now I terminate the loop after locating the first key